flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/SubpartitionCachedBuffer.java

```diff
@@ -0,0 +1,193 @@
+
+    private void writeRecord(ByteBuffer record, Buffer.DataType dataType) {
+        checkArgument(!dataType.isEvent());
+
+        ensureCapacityForRecord(record);
+
+        writeRecord(record);
+    }
+
+    private void ensureCapacityForRecord(ByteBuffer record) {
+        final int numRecordBytes = record.remaining();
+        int availableBytes =
+                Optional.ofNullable(unfinishedBuffers.peek())
+                        .map(
+                                currentWritingBuffer ->
+                                        currentWritingBuffer.getWritableBytes()
+                                                + bufferSize * (unfinishedBuffers.size() - 1))
+                        .orElse(0);
+
+        while (availableBytes < numRecordBytes) {
+            // request unfinished buffer.
+            BufferBuilder bufferBuilder = requestBufferFromPool();
+            unfinishedBuffers.add(bufferBuilder);
+            availableBytes += bufferSize;
+        }
+    }
+
+    private void writeRecord(ByteBuffer record) {
+        while (record.hasRemaining()) {
+            BufferBuilder currentWritingBuffer =
+                    checkNotNull(
+                            unfinishedBuffers.peek(), "Expect enough capacity for the record.");
+            currentWritingBuffer.append(record);
+            if (currentWritingBuffer.isFull()) {
+                finishCurrentWritingBuffer();
+            }
+        }
+    }
+
+    private void finishCurrentWritingBufferIfNotEmpty() {
+        BufferBuilder currentWritingBuffer = unfinishedBuffers.peek();
+        if (currentWritingBuffer == null || currentWritingBuffer.getWritableBytes() == bufferSize) {
+            return;
+        }
+
+        finishCurrentWritingBuffer();
+    }
+
+    private void finishCurrentWritingBuffer() {
+        BufferBuilder currentWritingBuffer = unfinishedBuffers.poll();
+        if (currentWritingBuffer == null) {
+            return;
+        }
+        currentWritingBuffer.finish();
+        BufferConsumer bufferConsumer = currentWritingBuffer.createBufferConsumerFromBeginning();
+        Buffer buffer = bufferConsumer.build();
+        currentWritingBuffer.close();
+        bufferConsumer.close();
+        addFinishedBuffer(
+                new NetworkBuffer(
+                        buffer.getMemorySegment(),
+                        // Note that the buffer recycler should be replaced before writing to the
+                        // tiers.
+                        FreeingBufferRecycler.INSTANCE,
+                        buffer.getDataType(),
+                        buffer.getSize()));
+    }
+
+    private void addFinishedBuffer(Buffer finishedBuffer) {
+        bufferFlusher.accept(subpartitionId, Collections.singletonList(finishedBuffer));
+    }
+
+    public BufferBuilder requestBufferFromPool() {
+        MemorySegment segment = storageMemoryManager.requestBufferInAccumulator();
+        cacheFlushManager.checkNeedTriggerFlushCachedBuffers();
```

Comment:
Still don't understand what does this do.

Reply:
After requesting a new buffer from the memory manager, we should check whether it is needed to flush the cached buffers to release buffers. This may include flushing the cached buffers to disk or remote storage.